Code File (Custom Classes) Docs (#363)
FlutterFlow/flutterflow-documentation: docs/ff-concepts/adding-customization/code-file.md

:::

## Add Custom Class

To add a custom class, go to the **Custom Code** from the left navigation menu, click **plus (+)** button, and select **Code File**. Set the name of the file, add your code, and hit the **Save** button. Now, you must **validate** your code in the editor to catch basic syntax errors. If there are no errors, click the **Parse** button. FlutterFlow will scan your code and automatically detect supported classes and enums.

Here’s an example of adding a Review custom class:

<div style={{
    position: 'relative',
    paddingBottom: 'calc(56.67989417989418% + 41px)', // Keeps the aspect ratio and additional padding
    height: 0,
    width: '100%'}}>
    <iframe 
        src="https://demo.arcade.software/sX8BKGhdsUGj7IZJCSpN?embed&show_copy_link=true"
        title=""
        style={{
            position: 'absolute',
            top: 0,
            left: 0,
            width: '100%',
            height: '100%',
            colorScheme: 'light'
        }}
        frameborder="0"
        loading="lazy"
        webkitAllowFullScreen
        mozAllowFullScreen
        allowFullScreen
        allow="clipboard-write">
    </iframe>
</div>

> <p></p>

Author: Done!

## Create Class Instance

> You need to create an instance of a class so you can work with actual data and use the class’s properties and methods in your app. Here’s a simple explanation:

Reviewer: I guess we don't have any docs in custom data type about creating a datatype instance?

> You need to create an instance of a class so you can work with actual data and use the class’s properties and methods in your app. Here’s a simple explanation:

Author: yes, we don't have it yet.

> You need to create an instance of a class so you can work with actual data and use the class’s properties and methods in your app. Here’s a simple explanation:

Reviewer: can you create a linear ticket for the same and reference this PR comment?

> You need to create an instance of a class so you can work with actual data and use the class’s properties and methods in your app. Here’s a simple explanation:

Author: Created - https://linear.app/flutterflow/issue/DEVR-945/add-docs-in-custom-data-type-about-creating-a-datatype-instance

- A **class** is like a blueprint or template. For example, the `Review` class describes what a review is, but doesn’t hold any real review information itself.
- An **instance** (or “object”) is a real, usable item made from that blueprint. See the code snippet below:
    
    ```jsx
    Review review1 = Review(
      'r001',
      'p123',
      'u456',
      'Alex Morgan',
      'Great quality T-shirt!',
      4.5,
      DateTime(2025, 5, 22),
      3,
      ReviewStatus.approved,
    );
    ```
    
- You can reuse the same class structure multiple times with different review data.

When you create an instance of a class, you can:

- Store actual review details.
- Access and update the fields (e.g., `review1.rating` or `review1.comment`).

> - Call methods that do something with that data (e.g., `review1.markHelpful()` or `review1.shortComment()`).

Author: Ya, now worries. added the code snippet for the Review class.

To create an instance of a custom class, open the **Set from Variable** dialog and select **Create Custom Class Instance**. Choose the class you want to use, then select the class name from the **Constructor** dropdown. After that, set values for each of the required fields.

<div style={{
    position: 'relative',
    paddingBottom: 'calc(56.67989417989418% + 41px)', // Keeps the aspect ratio and additional padding
    height: 0,
    width: '100%'}}>
    <iframe 
        src="https://demo.arcade.software/0CQju0ZUuhOEnkRhbxP4?embed&show_copy_link=true"
        title=""
        style={{
            position: 'absolute',
            top: 0,
            left: 0,
            width: '100%',
            height: '100%',
            colorScheme: 'light'
        }}
        frameborder="0"
        loading="lazy"
        webkitAllowFullScreen
        mozAllowFullScreen
        allowFullScreen
        allow="clipboard-write">
    </iframe>
</div>
<p></p>

> :::tip[When You Don't Need an Instance]

Reviewer: i see a bunch of new actions under custom class, see here: https://linear.app/flutterflow/issue/SUP-2529/custom-class-method-is-not-appearing-in-actions-list#comment-34cd1a72
also some related to static fields. I think it should be covered in the v1 version of the docs (as a separate section)?

> :::tip[When You Don't Need an Instance]

Author: When you say this '_should be covered in the v1 version of the docs_' you mean in this PR only right? yes, we can do it. Could you please help with that [ticket](https://linear.app/flutterflow/issue/SUP-2529/custom-class-method-is-not-appearing-in-actions-list#comment-34cd1a72)? it is is not opening for me.
<img width="796" alt="Screenshot 2025-05-27 at 5 18 17 PM" src="https://github.com/user-attachments/assets/b43d6113-53df-4380-9512-73bf31f106bd" />

> :::tip[When You Don't Need an Instance]

Author: Done! Added a new 'Using Static Classes' section.

You can skip instance creation when everything in the class is `static`, meaning it's shared across all uses (like a utility class). For example, look at the class below:

```jsx
class Utils {
  static int square(int x) => x * x;
}
```

In such cases, you can directly access the class data and methods via the **Set from Variable** menu.

![static-class-methods.avif](imgs/static-class-methods.avif)
:::

## Using Custom Class

Once the custom class is added successfully, you can access its fields and methods in the Variable Dialog, call its methods in the Action Flow Editor, assign instances to state variables, pass them to page or component parameters, and use enum values in dropdowns or conditionals.

### Data Types

You can select your custom class as a Type for variables, state, or parameters, just like a [Custom Data Type](../../resources/data-representation/custom-data-types.md).

![custom-class-as-data-type.avif](imgs/custom-class-as-data-type.avif)

### Access Fields and Methods

You can use custom class fields to display values directly in the UI, and call its methods in variable dialogs to return a result.

![access-fields-methods.avif](imgs/access-fields-methods.avif)

> ### Call In Action Flows

Author: Yes, you're right. However, there's currently an [issue](https://linear.app/flutterflow/issue/ENG-8084/cutsom-class-method-is-not-visible-in-actions-list#comment-f23808fa) where custom class methods aren't showing up in the actions list, so I’m unable to see any details or record the arcade.

![custom-enums.avif](imgs/custom-enums.avif)

You can access the custom enums from **Set from Variable** menu > **Custom Enum** section. You’ll see your Dart file listed by name. Select the enum you want to use, such as `ReviewStatus`, and then choose the specific value you want to assign.

<div style={{
    position: 'relative',
    paddingBottom: 'calc(56.67989417989418% + 41px)', // Keeps the aspect ratio and additional padding
    height: 0,
    width: '100%'}}>
    <iframe 
        src="https://demo.arcade.software/wFd9NKYGdb2Jp7ZYkaR8?embed&show_copy_link=true"
        title=""
        style={{
            position: 'absolute',
            top: 0,
            left: 0,
            width: '100%',
            height: '100%',
            colorScheme: 'light'
        }}
        frameborder="0"
        loading="lazy"
        webkitAllowFullScreen
        mozAllowFullScreen
        allowFullScreen
        allow="clipboard-write">
    </iframe>
</div>
<p></p>

## Tips & Best Practices

- Keep your custom class files modular and focused; ideally one class per file for better organization and reusability.

> - Avoid advanced Dart features that are not supported by FlutterFlow’s parser, such as generics or function-typed fields.

Author: I believe that if it isn't parsed, it won't be available in FlutterFlow's variable dialog or Action Flow editor.

As per wenkai's comment [here](https://linear.app/flutterflow/issue/DEVR-914/custom-classes):

> Certain things are not supported (class with generics, field/method arguments that is of function type, extensions, etc). They will be ignored.The parsed result contains only things FF can understand/use on the frontend.

            position: 'absolute',
            top: 0,
            left: 0,
            width: '100%',
            height: '100%',
            colorScheme: 'light'
        }}
        frameborder="0"
        loading="lazy"
        webkitAllowFullScreen
        mozAllowFullScreen
        allowFullScreen
        allow="clipboard-write">
    </iframe>
</div>
<p></p>

## Using Custom Class

Once the custom class is added successfully, you can access its fields and methods in the Variable Dialog, call its methods in the Action Flow Editor, assign instances to state variables, pass them to page or component parameters, and use enum values in dropdowns or conditionals.

### Custom Class as Data Type

You can select your custom class as a Type for variables, state, or parameters, just like a [Custom Data Type](../../resources/data-representation/custom-data-types.md).

![custom-class-as-data-type.avif](imgs/custom-class-as-data-type.avif)

### Access Fields and Methods

You can use custom class fields to display values directly in the UI, and call its methods in variable dialogs to return a result.

![access-fields-methods.avif](imgs/access-fields-methods.avif)

### Set Field [Action]

> Use the **Set Field** action to update a specific property of a custom class instance. For example, you can set `order.status = 'shipped'` when a user confirms an order, allowing the UI to reflect the change immediately.

Reviewer: The example is different from the text here. May be confusing to users.

Reviewer: + not matching the Review class above

            colorScheme: 'light'
        }}
        frameborder="0"
        loading="lazy"
        webkitAllowFullScreen
        mozAllowFullScreen
        allowFullScreen
        allow="clipboard-write">
    </iframe>
</div>
<p></p>

### Call Method [Action]

Use the **Call Method** action to invoke a method defined in your custom class. For instance, if your `Comment` class has a `markHelpful()` method, you can trigger it when a user taps a “Helpful” button to record the interaction.

<div style={{
    position: 'relative',
    paddingBottom: 'calc(56.67989417989418% + 41px)', // Keeps the aspect ratio and additional padding
    height: 0,
    width: '100%'}}>
    <iframe 
        src="https://demo.arcade.software/dVK7pbda7U6B4igmNUP7?embed&show_copy_link=true"
        title=""
        style={{
            position: 'absolute',
            top: 0,
            left: 0,
            width: '100%',
            height: '100%',
            colorScheme: 'light'
        }}
        frameborder="0"
        loading="lazy"
        webkitAllowFullScreen
        mozAllowFullScreen
        allowFullScreen
        allow="clipboard-write">
    </iframe>
</div>
<p></p>

> ## Using Static Classes

Reviewer: I'm confused on this one. technically Dart doesnt allow a "static class" per se. What people often do instead is create a class that contains only static members or mix static and instance members in the same class. So the title might be a bit misleading in this context? Maybe just call it Using Static Members if you agree?

Author: ok updated to 'Using Static Members'

> Sometimes, you just want to run a small piece of logic like formatting text or doing a calculation, without needing to set up anything extra. That’s when `static` data members are helpful. You can access static fields and methods without creating an instance of a class. You can think of them like tools that are always ready to use.

Reviewer: This is incorrect actually. We don't use static classes because its minor logic, but rather when there is "shared logic across app" where calculator / utility functions fit that definition.

Reviewer: ```suggestion
Sometimes, you just want to run a small piece of logic like formatting text or doing a calculation, without needing to set up anything extra. That’s when `static` data members are helpful. You can access static fields and methods without creating an instance of a class. Because they're tied to the class rather than an instance, static members are accessible globally.
```

This approach is typically used for **stateless utility classes** where shared functionality is needed across the app. For example, look at the class below:

```jsx
class Utils {
  static int square(int x) => x * x;
}
```

The `Utils` class contains a static method `square` that returns the square of a number without needing to create an object of the class.

Here are couple more examples to understand it better:

- This `StringFormatter` class below provides reusable static methods to capitalize text, convert it to lowercase, or format it in snake_case.

  ```jsx

>   class StringFormatter {

Reviewer: indentation is off on this one

Author: Updated!

  static String lastFormatted = '';
  static int formatCount = 0;
  
  static String capitalize(String input) =>
      input[0].toUpperCase() + input.substring(1);

  static String toLowerCase(String input) => input.toLowerCase();

  static String toSnakeCase(String input) =>
      input.replaceAll(' ', '_').toLowerCase();
  }
  ```

- The `MathHelper` class offers handy static methods to calculate tax, apply discounts, find percentages, and round off numbers.

  ```jsx
  class MathHelper {
    static double calculateTax(double amount) => amount * 0.18;

    static double applyDiscount(double amount, double discountPercent) =>
        amount - (amount * discountPercent / 100);

    static double calculatePercentage(double part, double total) =>
        (part / total) * 100;

    static int roundOff(double value) => value.round();
  }
  ```

:::warning

> Static classes are powerful, but they should be used carefully. Overusing static methods can lead to less flexible code and potential issues, especially when the logic requires access to state or needs to evolve over time. Stick to static methods only when the logic is truly independent and doesn’t rely on instance-specific data.

Reviewer: can fix the first line based on the feedback about "static classes"

Reviewer: also mention somewhere that static and instance fields can be mixed in a class as well?

You can access the static class data and methods directly via the ****Set from Variable**** menu.

![static-class-methods.avif](imgs/static-class-methods.avif)

### Set Static Field [Action]

Use the **Set Static Field** action to update a static field on a custom class. For example, if you have a class `MathHelper` with a static field `amount`, you can set it using an input value when a user enters a price. This allows you to store that value globally and use it across different calculations.

<div style={{
    position: 'relative',
    paddingBottom: 'calc(56.67989417989418% + 41px)', // Keeps the aspect ratio and additional padding
    height: 0,
    width: '100%'}}>
    <iframe 
        src="https://demo.arcade.software/TcYX7f2OgBZaayKFsKhN?embed&show_copy_link=true"
        title=""
        style={{
            position: 'absolute',
            top: 0,
            left: 0,
            width: '100%',
            height: '100%',
            colorScheme: 'light'
        }}
        frameborder="0"
        loading="lazy"
        webkitAllowFullScreen
        mozAllowFullScreen
        allowFullScreen
        allow="clipboard-write">
    </iframe>
</div>
<p></p>

> ### Call Static Method [Action]

Reviewer: The example in the text not matching the example in the Arcade. You can add multiple examples, but the main one should reflect the Arcade

> ### Call In Action Flows

Reviewer: I dont think we need this, lets just have all actions in h3 and remove this.

Author: Done!
